src/game/common/rts/player.cpp

```diff
@@ -426,3 +986,191 @@ void Player::Set_Player_Type(PlayerType t, bool is_skirmish)
     Call_Method<void, Player, PlayerType, bool>(PICK_ADDRESS(0x00451050, 0x0085ABEC), this, t, is_skirmish);
 #endif
 }
+
+Upgrade *Player::Add_Upgrade(const UpgradeTemplate *upgrade_template, UpgradeStatusType status)
+{
+    Upgrade *upgrade = Find_Upgrade(upgrade_template);
+
+    if (upgrade == nullptr) {
+        upgrade = new Upgrade(upgrade_template);
+        upgrade->Friend_Set_Prev(nullptr);
+        upgrade->Friend_Set_Next(m_upgradeList);
+
+        if (m_upgradeList != nullptr) {
+            m_upgradeList->Friend_Set_Prev(upgrade);
+        }
+
+        m_upgradeList = upgrade;
+    }
+
+    upgrade->Set_Status(status);
+    BitFlags<128> mask = upgrade_template->Get_Upgrade_Mask();
+
+    if (status == UPGRADE_STATUS_IN_PRODUCTION) {
+        m_upgradesInProgress.Set(mask);
+    } else if (status == UPGRADE_STATUS_COMPLETE) {
+        m_upgradesInProgress.Clear(mask);
+        m_upgradesCompleted.Set(mask);
+        On_Upgrade_Completed(upgrade_template);
+    }
+
+    if (g_thePlayerList->Get_Local_Player() == this) {
+        g_theControlBar->Mark_UI_Dirty();
+    }
+
+    return upgrade;
+}
+
+Upgrade *Player::Find_Upgrade(const UpgradeTemplate *upgrade_template)
+{
+    for (Upgrade *upgrade = m_upgradeList; upgrade != nullptr; upgrade = upgrade->Friend_Get_Next()) {
+        if (upgrade->Get_Template() == upgrade_template) {
+            return upgrade;
+        }
+    }
+
+    return nullptr;
+}
+
+void Player::On_Upgrade_Completed(const UpgradeTemplate *upgrade_template)
+{
+    for (auto i = m_playerTeamPrototypes.begin(); i != m_playerTeamPrototypes.end(); i++) {
+        DLINK_ITERATOR<Team> iter = (*i)->Iterate_Team_Instance_List();
+
+        while (!iter.Done()) {
+            Team *team = iter.Cur();
+
+            if (team != nullptr) {
+                DLINK_ITERATOR<Object> iter2 = team->Iterate_Team_Member_List();
+
+                while (!iter2.Done()) {
+                    Object *obj = iter2.Cur();
+
+                    if (obj != nullptr) {
+                        obj->Update_Upgrade_Modules();
+                    }
+
+                    iter2.Advance();
+                }
+            }
+
+            iter.Advance();
+        }
+    }
+}
+
+void Player::Delete_Upgrade_List()
+{
+    while (m_upgradeList != nullptr) {
+        Upgrade *upgrade = m_upgradeList->Friend_Get_Next();
+        m_upgradeList->Delete_Instance();
+        m_upgradeList = upgrade;
+    }
+
+    m_upgradesInProgress.Clear();
+    m_upgradesCompleted.Clear();
+}
+
+void Player::Reset_Rank()
+{
+    m_rankLevel = 1;
+    m_currentSkillPoints = 0;
+    const RankInfo *info = g_theRankInfoStore->Get_Rank_Info(m_rankLevel + 1);
+
+    if (info != nullptr) {
+        m_skillPointsNeededForNextRank = info->m_skillPointsNeeded;
```

Comment:
This member variable remains uninitialized if there is no next rank configured.

Reply:
fixed.